Implement static code analysis rule to detect vulnerable environment variable initializations

nureka-rodrigo opened the pull request:
## Purpose

Fixes https://github.com/ballerina-platform/ballerina-library/issues/7990

## Approach

1. When a function call is encountered, the analyzer checks if it's calling `os:setEnv` using the `isOsSetEnvCall` method.
2. If an `setEnv` call is found, the analyzer calls `containsUntrustedEnvValue` to check the arguments:
   - It examines each argument in the function call
   - It checks specifically for the value parameter (either as a named argument or as the second positional argument)
   - It determines if the value comes from user-controlled input
3. If untrusted input is detected, the analyzer reports an issue

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility

nureka-rodrigo committed:
Implement static code analysis rule to detect vulnerable environment variable initializations

ThisaruGuruge commented on compiler-plugin-tests/src/test/java/io/ballerina/stdlib/os/compiler/staticcodeanalyzer/StaticCodeAnalyzerTest.java:
```diff
@@ -1,116 +1,137 @@
+            Assertions.assertRule(
+                    rules,
+                    "ballerina/os:1",
+                    "Avoid constructing system command arguments from user input without proper sanitization",
```
Shall we use constants for these values instead of using `String` values?

nureka-rodrigo replied:
Fix: b622d8ead9ce0679c39e35fb36e367563df0889d

ThisaruGuruge commented on compiler-plugin-tests/src/test/java/io/ballerina/stdlib/os/compiler/staticcodeanalyzer/StaticCodeAnalyzerTest.java:
```diff
@@ -1,116 +1,137 @@
+                Assert.assertEquals(issues.size(), 1);
+                Assertions.assertIssue(issues, index, "ballerina/os:2", "main.bal",
+                        20, 23, Source.BUILT_IN);
+            }
```
Using an `if-else` block inside the loop to differentiate loop members is bad practice. Shall we use an indexed (while) loop instead and do the checking?

nureka-rodrigo replied:
Fix: 463c2b1b06564ffb277e96430741158a8282604d

ThisaruGuruge commented on compiler-plugin/src/main/java/io/ballerina/stdlib/os/compiler/staticcodeanalyzer/OSCommandInjectionAnalyzer.java:
```diff
@@ -92,23 +94,35 @@ public void perform(SyntaxNodeAnalysisContext context) {
+            return false;
+        }
+
+        return importPrefix.contains(qNode.modulePrefix().text()) && qNode.identifier().text().equals(SET_ENV);
```
Can't we use semantic API here instead of string comparison?

nureka-rodrigo replied:
Fix: 6790036465e846cb9ba84c61b6d7d93819550530

ThisaruGuruge commented on compiler-plugin/src/main/java/io/ballerina/stdlib/os/compiler/staticcodeanalyzer/OSCommandInjectionAnalyzer.java:
```diff
@@ -125,6 +139,27 @@ private boolean containsUserControlledInput(SeparatedNodeList<FunctionArgumentNo
+            }
+
+            if (((arg instanceof NamedArgumentNode named && named.argumentName().name().text().equals(VALUE))
+                    || (arg instanceof PositionalArgumentNode && idx == 1)) && isUserControlledInput(expr, context)) {
```
Would be easier to read if these are moved to a separate methods.

nureka-rodrigo replied:
Fix: 47f8b4c4feff0b3a4c61797073ac8987e501a4b2

ThisaruGuruge commented on compiler-plugin/src/main/java/io/ballerina/stdlib/os/compiler/staticcodeanalyzer/RuleFactory.java:
```diff
@@ -26,7 +26,8 @@
  */
 public class RuleFactory {
```
Make this class final?

nureka-rodrigo replied:
Fix: 532932dc944f68a58a6f3709004cf1057f2f2ca0

nureka-rodrigo committed:
Add getDescription method to OSRule and update tests to use dynamic descriptions

nureka-rodrigo committed:
Make RuleFactory class final

nureka-rodrigo committed:
Refactor OSCommandInjectionAnalyzer to improve argument checks with helper methods

nureka-rodrigo committed:
Refactor StaticCodeAnalyzerTest to improve test structure and readability

nureka-rodrigo committed:
Add Jackson Databind dependency and refactor JSON normalization in StaticCodeAnalyzerTest

ThisaruGuruge commented on compiler-plugin-tests/src/test/java/io/ballerina/stdlib/os/compiler/staticcodeanalyzer/StaticCodeAnalyzerTest.java:
```diff
@@ -1,126 +1,170 @@
+                        20, 23, Source.BUILT_IN);
+                break;
+            default:
+                break;
```
Shouldn't we `assertFail` here?

nureka-rodrigo replied:
Fix: c5e2fdc6d29e246f8bbfde6fa6e2b717379d7dca

nureka-rodrigo committed:
Add failure assertion for unhandled rules in StaticCodeAnalyzerTest

nureka-rodrigo committed:
Refactor file paths in static analysis rules to be relative

Updated the file paths in rule<x>.json to use relative paths instead of absolute paths. This change improves portability and ensures that the rules can be used in different environments without modification.

nureka-rodrigo committed:
Refactor OSCommandInjectionAnalyzer to enhance function call checks

Updated the OSCommandInjectionAnalyzer to utilize a shared helper method for validating OS function calls. This change improves code readability and maintainability by reducing redundancy in the isOsExecCall and isOsSetEnvCall methods. Additionally, integrated semantic model checks to ensure accurate function and module prefix validation.

nureka-rodrigo commented:
@ThisaruGuruge PRAM
